helix-term/src/ui/editor.rs

```diff
@@ -145,11 +145,20 @@ impl EditorView {
             highlights = Box::new(syntax::merge(highlights, overlay_highlights));
         }
 
-        // update document-wide search match highlights
-        highlights = match view.get_search_matches(theme) {
-            Some(match_highlights) => Box::new(syntax::merge(highlights, match_highlights)),
-            None => highlights,
-        };
+        // update document-wide search match highlights, but only if the
+        // document remains unchanged
+        if doc.changes().is_empty() {
+            highlights = match view.get_search_matches(theme) {
+                Some(match_highlights) => Box::new(syntax::merge(highlights, match_highlights)),
+                None => highlights,
+            };
+        } else {
+            // if document has changed, set bool in view
+            // view.search_matches.need_clear.replace(true);
+            view.search_matches
+                .need_clear
+                .swap(true, std::sync::atomic::Ordering::Relaxed);
+        }
```

Comment:
this is where i 'clear' the highlights if the doc has been modified. `if doc.changes().is_empty()` then i update the highlights, otherwise, i use my insanly convoluted `AtomicBool` to tell View to clear out `search_matches` as soon as i exit the current routine, and get back to a scope where view is mutable.